src/Symfony/Component/HttpClient/CurlHttpClient.php

```diff
@@ -0,0 +1,314 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Symfony\Component\HttpClient\Exception\InvalidArgumentException;
+use Symfony\Component\HttpClient\Exception\TransportException;
+use Symfony\Component\HttpClient\Response\CurlResponse;
+use Symfony\Component\HttpClient\Response\ResponseStream;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\HttpClientTrait;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+
+/**
+ * A performant implementation of the HttpClientInterface contracts based on the curl extension.
+ *
+ * This provides fully concurrent HTTP requests, with transparent
+ * HTTP/2 push when a curl version that supports it is installed.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class CurlHttpClient implements HttpClientInterface
+{
+    use HttpClientTrait;
+
+    private $defaultOptions = self::OPTIONS_DEFAULTS;
+    private $noProxy;
+    private $multi;
+
+    /**
+     * @param array $defaultOptions     Default requests' options
+     * @param int   $maxHostConnections The maximum number of connections to a single host
+     * @param int   $noProxy            A comma separated list of hosts that do not require a proxy to be reached, even if one is specified
+     */
+    public function __construct(array $defaultOptions = [], int $maxHostConnections = 6, string $noProxy = null)
+    {
+        if ($defaultOptions) {
+            [, $this->defaultOptions] = self::prepareRequest(null, null, $defaultOptions, self::OPTIONS_DEFAULTS);
+        }
+
+        $mh = curl_multi_init();
+
+        // Don't enable HTTP/1.1 pipelining: it forces responses to be sent in order
+        curl_multi_setopt($mh, CURLMOPT_PIPELINING, CURLPIPE_MULTIPLEX);
+        curl_multi_setopt($mh, CURLMOPT_MAX_HOST_CONNECTIONS, $maxHostConnections);
+
+        $this->noProxy = $noProxy;
+
+        // Use an internal stdClass object to share state between the client and its responses
+        $this->multi = $multi = (object) [
+            'openHandles' => [],
+            'handlesActivity' => [],
+            'handle' => $mh,
+            'pushedResponses' => [],
+            'dnsCache' => [[], [], []],
+        ];
+
+        // Skip configuring HTTP/2 push when it's unsupported or buggy, see https://bugs.php.net/76675
+        if (\PHP_VERSION_ID < 70215 || \PHP_VERSION_ID === 70300 || \PHP_VERSION_ID === 70301) {
+            return;
+        }
+
+        // HTTP/2 push crashes before curl 7.61
+        if (0x073d00 > ($v = curl_version())['version_number'] || !(CURL_VERSION_HTTP2 & $v['features'])) {
+            return;
+        }
+
+        // Keep a dummy "onPush" reference to work around a refcount bug in PHP
+        curl_multi_setopt($mh, CURLMOPT_PUSHFUNCTION, $multi->onPush = static function ($parent, $pushed, array $rawHeaders) use ($multi) {
```

Comment:
`: int` ?

Reply:
@nicolas-grekas Most anonymous functions are missing a strict return type, is it on purpose?

Reply:
yes: they're internal, we don't care